pandas/core/series.py

```diff
@@ -3307,21 +3284,24 @@ def _try_kind_sort(arr):
         if not is_bool(ascending):
             raise ValueError("ascending must be boolean")
 
-        if not ascending:
-            argsorted = argsorted[::-1]
-
-        if na_position == "last":
-            n = good.sum()
-            sorted_index[:n] = idx[good][argsorted]
-            sorted_index[n:] = idx[bad]
-        elif na_position == "first":
-            n = bad.sum()
-            sorted_index[n:] = idx[good][argsorted]
-            sorted_index[:n] = idx[bad]
-        else:
-            raise ValueError(f"invalid na_position: {na_position}")
+        arr = self._values
 
-        result = self._constructor(arr[sorted_index], index=self.index[sorted_index])
+        if key:
+            if isinstance(arr, SparseArray):
+                # SparseArray doesn't store NaNs item-by-item, so pass everything
+                arr = ensure_key_mapped(self, key)._values
+            else:
+                good = ~isna(arr)
+                keyed = ensure_key_mapped(self[good], key)._values
+                arr = arr.astype(keyed.dtype)
+                arr[good] = keyed
+
+        # GH 35922. Make sorting stable by leveraging nargsort
+        sorted_index = nargsort(arr, kind, ascending, na_position)
+
+        result = self._constructor(
+            self._values[sorted_index], index=self.index[sorted_index]
+        )
```

Comment:
Since we change `arr` when we sort with a key function, use original values here.